src/crimson/net/Protocol.h

```diff
@@ -34,11 +35,18 @@ class Protocol {
   // Reentrant closing
   void close(bool dispatch_reset, std::optional<std::function<void()>> f_accept_new=std::nullopt);
   seastar::future<> close_clean(bool dispatch_reset) {
-    close(dispatch_reset);
-    // it can happen if close_clean() is called inside Dispatcher::ms_handle_reset()
-    // which will otherwise result in deadlock
-    assert(close_ready.valid());
-    return close_ready.get_future();
+    // yield() so that close(dispatch_reset) can be called *after*
+    // close_clean() is applied to all connections in a container using
+    // seastar::parallel_for_each(). otherwise, we could erase a connection in
+    // the container when seastar::parallel_for_each() is still iterating in
+    // it. that'd lead to a segfault.
+    return seastar::yield().then([this, dispatch_reset] {
```

Comment:
@tchaikov so the yield will cause this continuation is  scheduled after the other tasks finished in the same schedule group, right?

Reply:
@liu-chunmei kinda. as the terminology of `scheduling_group` has its specific meaning in Seastar, i'd prefer describe the solution as: call `close()` after `seastar::parallel_for_each()`. as without this change, both the futurized continuations (i.e., the `close()` calls) created by `seastar::parallel_for_each()` and its body are executed in the same continutation if `SEASTAR_DEBUG` is not defined. after this change, the body of `seastar::parallel_for_each()` is executed first. only after all the futures are collected, the continuation pending on `yield()` are scheduled.